### PR TITLE
feat(refactor) - refactor the crypto package

### PR DESCRIPTION
**Pull request by ctrlc03**
Refactor the crypto package:
- [x] shift from one file to multiple organized files
- [x] add more typedoc docs
- [x] ensure generateRandomBabyJubValue does not throw but runs until a valid value is generated
- [x] separate types
- [x] separate tests and move supporting functions to utils
- [x] fix inconsistencies between BigInt and bigint types
- [x] fix various (repo-wide) ts bugs due to missing types declarations

### CODE REVIEW

File: crypto/ts/AccQueue.ts [hunk 1]
```diff
@@ -146,32 +122,46 @@ class AccQueue {
...
+        assert(this.numLeaves < this.hashLength ** this.MAX_DEPTH, "AccQueue is full")
 
-        // Ensure that _value is a BigInt
         this._enqueue(_leaf, 0)
 
+        // the index is the number of leaves (0-index)
         const leafIndex = this.numLeaves
 
-        this.numLeaves ++
+        // increase the number of leaves
```
Comment (baumstern) [inline crypto/ts/AccQueue.ts:150]: nice comments

File: crypto/ts/AccQueue.ts [hunk 2]
```diff
@@ -5,45 +5,21 @@ import {
...
-}
-
-interface Queue {
-    levels: BigInt[][];
-    indices: number[];
-}
+    IncrementalQuinTree
+} from './index'
+import {
+    Leaf,
```
Comment (baumstern) [inline crypto/ts/AccQueue.ts:11]: nice typing

File: crypto/ts/AccQueue.ts [hunk 3]
```diff
@@ -146,32 +122,46 @@ class AccQueue {
...
 
-        this.numLeaves ++
+        // increase the number of leaves
+        this.numLeaves++
+        // we set merged false because there are new leaves
         this.subTreesMerged = false
+        // reset the smallSRTroot because it is obsolete
         this.smallSRTroot = BigInt(0)
 
+        // @todo this can be moved in the constructor rather than computing every time
```
Comment (baumstern) [inline crypto/ts/AccQueue.ts:157]: +1

File: crypto/ts/__tests__/AccQueue.test.ts [hunk 4]
```diff
@@ -2,149 +2,7 @@ import {
...
-
-        const depth = calcDepthFromNumLeaves(HASH_LENGTH, numSubtrees)
-        const smallTree = new IncrementalQuinTree(depth, aq.zeros[aq.subDepth], HASH_LENGTH, aq.hashFunc)
-        for (let i = 0; i < aq.subRoots.length; i++) {
-            smallTree.insert(aq.subRoots[i])
-        }
-        expect(aq.smallSRTroot.toString()).toEqual(smallTree.root.toString())
-    }
-}
+import { testMerge, testMergeExhaustive, testMergeShortest, testMergeShortestOne } from './utils'
```
Comment (baumstern) [inline crypto/ts/__tests__/AccQueue.test.ts:5]: nice

**Comment by baumstern** [comment]
Great work on this! Just I need to address the CI issues

**Comment by baumstern** [comment]
Relevant PR of CI fix: https://github.com/privacy-scaling-explorations/maci/pull/775

**Comment by samajammin** [comment]
> fix inconsistencies between BigInt and bigint types

Just for my understanding, `bigint` is the type, `BigInt` is the constructor, is that correct?
https://stackoverflow.com/questions/76477290/what-is-the-difference-between-bigint-lowercase-and-bigint

**Comment by samajammin** [comment]
> add more typedoc docs

Noticing that `typedoc` is not installed. Is the thought for now to just add comments that support TypeDoc, then later install it to create the static doc site?

File: crypto/README.md [hunk 5]
```diff
@@ -90,29 +90,29 @@ Checks whether the given signature is valid.
 
 ### Hash functions
 
-**`sha256Hash = (input: BigInt[]): BigInt`**: a wrapper function over
+**`sha256Hash = (input: bigint[]): bigint`**: a wrapper function over
```
Comment (samajammin) [inline crypto/README.md:93]: Thoughts on having these explanations in the README? Seems like it may become redundant work with TypeDoc.

File: crypto/ts/utils.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+ * Create a copy of a bigint array
+ * @param arr - the array of bigints to copy
+ * @returns a deep copy of the array
+ */
+export const deepCopyBigIntArray = (arr: bigint[]): bigint[] => {
+    return arr.map((x) => BigInt(x.toString()))
+}
+
+/**
+ * Calcualte the depth of a tree given the number of leaves
```
Comment (samajammin) [inline crypto/ts/utils.ts:37]:
```suggestion
 * Calculate the depth of a tree given the number of leaves
```

File: circuits/ts/__tests__/Hasher.test.ts [hunk 7]
```diff
@@ -191,7 +191,7 @@ describe('Poseidon hash circuits', () => {
         const circuit = 'messageHasher_test'
         it('correctly hashes a message', async () => {
             const k = new Keypair()
-            const random50bitBigInt = (): BigInt => {
+            const random50bitBigInt = (): bigint => {
```
Comment (samajammin) [inline circuits/ts/__tests__/Hasher.test.ts:194]: For future reference, I'd argue this `BigInt` --> `bigint` refactor should be a single PR. Would make this more manageable to review.

File: circuits/ts/__tests__/TallyVotes.test.ts [hunk 8]
```diff
@@ -261,7 +261,7 @@ describe('TallyVotes circuit', () => {
             poll.messageAq.merge(treeDepths.messageTreeDepth)
 
             for (let i = 0; i < NUM_BATCHES; i ++) {
-                poll.processMessages()
+                poll.processMessages(pollId)
```
Comment (samajammin) [inline circuits/ts/__tests__/TallyVotes.test.ts:264]: Why this change? Seems unrelated to the PR

File: core/ts/MaciState.ts [hunk 9]
```diff
@@ -1563,34 +1559,13 @@ const genSubsidyVkSig = (
...
 
-/*
- * A helper function which hashes a list of results with a salt and returns the
- * hash.
- *
- * @param results A list of vote weights
- * @parm salt A random salt
- * @return The hash of the results and the salt, with the salt last
- */
-const genTallyResultCommitment = (
```
Comment (samajammin) [inline core/ts/MaciState.ts:1581]: Why are we deleting this function?

File: crypto/ts/utils.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        depth ++
+    }
+
+    return depth
+}
+
+/**
+ * A helper function which hashes a list of results with a salt and returns the
+ * hash.
+ * @param results A list of values
```
Comment (samajammin) [inline crypto/ts/utils.ts:61]:
```suggestion
 * @param leaves A list of values
```

File: core/ts/MaciState.ts (hunk 9 above)
Comment (samajammin) [reply core/ts/MaciState.ts:1581]: Ah, I see we moved it to `maci-crypto` & renamed `genTreeCommitment`.

**Comment by ctrlc03** [comment]
> > add more typedoc docs
> 
> Noticing that `typedoc` is not installed. Is the thought for now to just add comments that support TypeDoc, then later install it to create the static doc site?

installating and setting up typedoc it's in its own PR: https://github.com/privacy-scaling-explorations/maci/pull/755
And yes you are correct, we add comments on functions following this standard, and once this #755 is merged, we can generate the code before commits (or decide on CI/pre-commit hooks to do that)

File: crypto/README.md (hunk 5 above)
Comment (ctrlc03) [reply crypto/README.md:93]: Good point! I'll amend the Readme

File: crypto/ts/utils.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ * @parm salt A random salt
+ * @returns The hash of the leavees and the salt, with the salt last
+ */
+export const genTreeCommitment = (
```
Comment (samajammin) [inline crypto/ts/utils.ts:65]: Should this function have a test?

File: circuits/ts/__tests__/TallyVotes.test.ts (hunk 8 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/TallyVotes.test.ts:264]: Unfortunately, this repo is still at a stage where you touch one little thing, and 1000 things break. I'll give you an example, you remove a console.log from the cli, and all integration tests fail 😓 
In this case, I added a declaration on the tsconfig file to generate type declarations -> this was done on the root tsconfig (which every package tsconfig extends) thus affecting the whole repo -> result is that there are quite a few mistakes in the codebase which just work, but Typescript is now complaining due to the type declarations being added

File: core/ts/MaciState.ts (hunk 9 above)
Comment (ctrlc03) [reply core/ts/MaciState.ts:1581]: Correct, this function seems to better live in crypto and could also help with this: https://github.com/privacy-scaling-explorations/maci/issues/781

Comment (ctrlc03) [reply core/ts/MaciState.ts:1581]: Also the name was changed as imo a more general name is appropriate

File: crypto/ts/utils.ts (hunk 11 above)
Comment (ctrlc03) [reply crypto/ts/utils.ts:65]: Totally, great catch

File: circuits/ts/__tests__/Hasher.test.ts (hunk 7 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/Hasher.test.ts:194]: Sure, I can do the rest of the repo in one PR, would be great to have this merged first though

File: core/ts/MaciState.ts [hunk 12]
```diff
@@ -38,7 +38,6 @@ interface BatchSizes {
 }
 
 interface MaxValues {
-    maxUsers: number;
```
Comment (samajammin) [inline core/ts/MaciState.ts:41]: Why remove `maxUsers`?

File: core/ts/MaciState.ts [hunk 13]
```diff
@@ -1236,8 +1235,6 @@ class Poll {
                     Number(this.batchSizes.messageBatchSize.toString()),
             },
             {
-                maxUsers:
```
Comment (samajammin) [inline core/ts/MaciState.ts:1239]: Again, why remove?

File: integrationTests/ts/__tests__/suites.ts [hunk 14]
```diff
@@ -41,13 +41,7 @@ const executeSuite = async (data: any, expect: any) => {
...
         const coordinatorKeypair = new Keypair()
 
-        const maciState = new MaciState(
-            coordinatorKeypair,
-            config.constants.maci.stateTreeDepth,
-            config.constants.maci.messageTreeDepth,
-            config.constants.maci.voteOptionTreeDepth,
-            config.constants.maci.maxVoteOptions,
-        )
+        const maciState = new MaciState()
```
Comment (samajammin) [inline integrationTests/ts/__tests__/suites.ts:44]: Why no longer passing those arguments?

**Review by samajammin — COMMENTED** [review]
Nice refactor overall!

Left some questions on things I don't understand.

File: integrationTests/ts/__tests__/suites.ts (hunk 14 above)
Comment (ctrlc03) [reply integrationTests/ts/__tests__/suites.ts:44]: Because now that there are type declarations, TS complaints that we do not need them, if you look at the constructor of MaciState it does not take any.

File: core/ts/MaciState.ts (hunk 12 above)
Comment (ctrlc03) [reply core/ts/MaciState.ts:41]: another type error + this is just not a valid value to insert there. The max number of users is determined by the depth of the state tree; also this maci-core is supposed to mirror the smart contracts, and in the smart contracts the MaxValues struct does not have maxUsers

File: core/ts/MaciState.ts [hunk 15]
```diff
@@ -1563,34 +1559,13 @@ const genSubsidyVkSig = (
...
            (BigInt(_intStateTreeDepth) << BigInt(64)) +
             BigInt(_voteOptionTreeDepth)
 }
 
 
-/*
- * A helper function which hashes a list of results with a salt and returns the
- * hash.
- *
- * @param results A list of vote weights
```
Comment (baumstern) [inline core/ts/MaciState.ts:1577]: Think we should retain this description somewhere.  
It's currently hard to understand what 'tallyFileData.perVOSpentVoiceCredits.tally.map((x) => BigInt(x))' means like following function call:
```
genTreeCommitment(
        tallyFileData.perVOSpentVoiceCredits.tally.map((x) => BigInt(x)),
        BigInt(tallyFileData.perVOSpentVoiceCredits.salt),
        poll.treeDepths.voteOptionTreeDepth)
```

Comment (ctrlc03) [reply core/ts/MaciState.ts:1577]: isn't the same (aside changing the word results to leaves) description on the genTreeCommitment function inside crypto/ts/utils.ts?

Comment (baumstern) [reply core/ts/MaciState.ts:1577]: Not sure if someone could figure out `tallyFileData.perVOSpentVoiceCredits.tally.map((x) => BigInt(x)` means a lists of 'vote weights'. 
Think we can separate the issue of 'documenting how to compute  `newResultsCommitment`, `newSpentVoiceCreditsCommitment`...etc in: https://github.com/privacy-scaling-explorations/maci/blob/c1b8b75698ed920f755fa1765e1a168096655fb3/cli/ts/genProofs.ts#L547-L557

**Review by baumstern — APPROVED** [review]
LGTM

File: circuits/ts/__tests__/TallyVotes.test.ts (hunk 8 above)
Comment (samajammin) [reply circuits/ts/__tests__/TallyVotes.test.ts:248]: Got it, appreciate the explanation!

File: core/ts/MaciState.ts (hunk 12 above)
Comment (samajammin) [reply core/ts/MaciState.ts:41]: Gotcha, thanks!

**Review by samajammin — APPROVED** [review]
Nice work 🚀
